src/editmap.cpp

```diff
@@ -1831,6 +1831,170 @@ int editmap::mapgen_preview( real_coords &tc, uimenu &gmenu )
+            }
+
+            destsm->field_count = srcsm->field_count; // and count
+            std::memcpy( destsm->trp, srcsm->trp, sizeof( srcsm->trp ) ); // traps
+            std::memcpy( destsm->rad, srcsm->rad, sizeof( srcsm->rad ) ); // radiation
+            std::memcpy( destsm->lum, srcsm->lum, sizeof( srcsm->lum ) ); // emissive items
+
+
+            for( int sx = 0; sx < SEEX; ++sx ) {
+                for( int sy = 0; sy < SEEY; ++sy ) {
+                    for( auto &elem : srcsm->itm[sx][sy] ) {
+                        destsm->itm[sx][sy].push_back(elem);
+                    }
+                    //Don't cover existing crops, for farm upgrades
+                    //Don't destroy terrain to place grass/dirt
+                    if( change_sensitive && destsm->frn[sx][sy] != furn_str_id( "f_plant_seed" ) &&
+                        destsm->frn[sx][sy] != furn_str_id( "f_plant_seedling" ) &&
+                        destsm->frn[sx][sy] != furn_str_id( "f_plant_mature" ) &&
+                        destsm->frn[sx][sy] != furn_str_id( "f_plant_harvest" ) &&
+                        srcsm->ter[sx][sy] != ter_str_id( "t_grass" ) &&
+                        srcsm->ter[sx][sy] != ter_str_id( "t_dirt" ) ) {
+                               destsm->ter[sx][sy] = srcsm->ter[sx][sy];
+                               destsm->frn[sx][sy] = srcsm->frn[sx][sy];
+                       }
+                    destsm->fld[sx][sy] = srcsm->fld[sx][sy];
+                    destsm->cosmetics[sx][sy].swap( srcsm->cosmetics[sx][sy] );
+                }
+            }
+
+            // various misc variables
+            destsm->active_items = srcsm->active_items;
+
+            destsm->temperature = srcsm->temperature;
+            destsm->last_touched = calendar::turn;
+            destsm->comp = std::move( srcsm->comp );
+            destsm->camp = srcsm->camp;
+
+            if( spawns_todo > 0 ) {
+                g->m.spawn_monsters( true );
+            }
+        }
+    }
+    g->m.reset_vehicle_cache( target.z );
+
+    cleartmpmap( tmpmap );
+    return true;
+}
+
+vehicle* editmap::mapgen_veh_query( tripoint omt_tgt, vehicle* car_target, bool destroy)
+{
+    tinymap target_bay;
+    target_bay.load( omt_tgt.x * 2, omt_tgt.y * 2, omt_tgt.z, false );
+
+    std::vector<vehicle *> possible_vehicles;
+    std::vector<vehicle *> sub_possible_vehicles;
+    for( int x = 0; x < 2; x++ ) {
+        for( int y = 0; y < 2; y++ ) {
+            submap *destsm = target_bay.get_submap_at_grid( x, y, target.z );
+            int target_pos = -1;
+            for( size_t z = 0; z < destsm->vehicles.size(); z++ ){
+                possible_vehicles.push_back(destsm->vehicles[z]);
+                if( destsm->vehicles[z] == car_target && destroy){
+                    target_pos = z;
+                    auto &ch = g->m.access_cache( destsm->vehicles[z]->smz );
+                    ch.vehicle_list.erase( destsm->vehicles[z] );
```

Comment:
You shouldn't remove things from a cache by hand. Remove the vehicle from the map and force the cache to regenerate.

Reply:
target_bay.destroy_vehicle(car_target);
g->m.destroy_vehicle(car_target);
target_bay.reset_vehicle_cache(target.z);
g->m.reset_vehicle_cache(target.z);

Attempting to destroy the vehicle from the target_bay crashes the game, when run through g->m the vehicle can't be found.  No combination of the above worked so I used the process in the existing method above mapgen_set(), mapgen_preview().  No idea why it crashes but there are a number of functions that I attempted to utilized that seem to not work as intended.  Like overmap_buffer.is_safe(), which I'm guessing now checks for the existence of hordes in a tile instead of hostile monster groups or ungenerated hostile monster groups?  I could never get it to return false no matter how dangerous the map tile.  Just trying to get this to the point that anyone can make optimizations if they know how.

Reply:
Your solution would result in random and hard to track crashes, so it would be much better if we found why `destroy` doesn't work.

The variant with `target_bay` should work. It does basically all you do here when `destroy == true`, but with extra sanity checks and cleanup.

Reply:
Oh I get it now.
You're erasing vehicles from their pointer lists, but aren't actually destroying them.
That's why using vehicles in `possible_vehicles` doesn't crash even after you delete those vehicles.

You have to take ownership of the deleted vehicle pointer at some point, otherwise the vehicle object will not be properly deleted.

Looking at your use case (taking a vehicle, operating on it, and then destroying it), it should work if you did this:
* Found the vehicles
* Picked one that you want
* Detached it (`target_bay.detach_vehicle`)
* Reloaded CURRENT map's (`g->m`) vehicle cache
* Cleared and re-filled current map's vehicle list. Note: there is currently no good method of doing the latter, you'd have to manually go through all submaps of current map and add the vehicles. See `map::loadn`.
* Now that vehicle is properly detached from the map, you can operate on the detached pointer
* Do NOT delete it afterwards - detached pointer will clear itself automatically when you exit the function without passing it

That part about vehicles in lists unfortunately has to be done. The only way to avoid it is to either only operate on vehicles on current map, only operate on vehicles outside current map, or have two code paths for vehicles on current map and outside it.

Reply:
So I got detach_vehicle to work after separating it into a separate method and throwing clear cache and vehicle list after it to ensure all traces of the pointer were gone.  I experimented with setting loadn to public and calling it both in/out of the targeted tile, tried the simple add_vehicle_to_cache that is used everywhere else but for some reason vehicles still didn't get reloaded.  

Noticed that I don't have to modify the vehicles, if you query for what vehicles are in the submap they get dropped from the cache so that they disappear until your movement reloads the cache???